File: src/two_numbers.rs
use std::collections::HashMap;

pub fn two_numbers_hashmap(array: &[isize], sum: isize) -> (isize, isize) {
    let mut acc = HashMap::new();

    for num in array.iter() {
        let summand = sum - num;

        match acc.get(&summand) {
            None => acc.insert(summand, true),
            Some(_value) => return (*num, summand),
        };
    }

    (0, 0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn is_right_two_numbers_found() {
        assert_eq!(
            two_numbers_hashmap(&[3, 5, -4, 8, 11, 11, -1, 6], 10),
            (11, -1)
        )
    }
}

File: typescript/two_numbers.ts
/**
 * complexity O(n)
 * space O(n)
 *
 * @param array
 * @param sum
 */
function two_numbers(array: number[], sum: number): number[] {
    const acc: { [key: string]: boolean } = {};

    for (let i = 0; i < array.length; i++) {
        const summand = sum - array[i];

        if (acc[array[i]]) {
            return [summand, array[i]]
        }

        acc[summand] = true
    }

    return [0, 0]
}


console.log('two_numbers:', two_numbers([3, 5, -4, 8, 11, 11, -1, 6], 10))
console.log('two_numbers:', two_numbers([3, -5, 4, 8, 13, 12, -1, 6], 10))
